src/java.base/share/classes/jdk/internal/classfile/impl/EntryMap.java

```diff
@@ -0,0 +1,194 @@
+
+        int capacity = arraySize(size, fillFactor);
+        this.fillFactor = fillFactor;
+        this.resizeThreshold = (int) (capacity * fillFactor);
+        this.mask1 = capacity - 1;
+        this.mask2 = capacity * 2 - 1;
+        data = new int[capacity * 2];
+    }
+
+    protected abstract T fetchElement(int index);
+
+    public int firstToken(int hash) {
+        if (hash == 0)
+            throw new IllegalArgumentException("hash must be nonzero");
+
+        int ix = (hash & mask1) << 1;
+        int k = data[ix];
+
+        if (k == 0)
+            return NO_VALUE;  //end of chain already
+        else if (k == hash)
+            return ix;
+        else
+            return nextToken(hash, ix);
+    }
+
+    public int nextToken(int hash, int token) {
+        int ix = token;
+        while (true) {
+            ix = (ix + 2) & mask2; // next index
+            int k = data[ix];
+            if (k == 0)
+                return NO_VALUE;
+            else if (k == hash)
+                return ix;
+        }
+    }
+
+    public int getIndexByToken(int token) {
+        return data[token + 1];
+    }
+
+    public T getElementByToken(int token) {
+        return fetchElement(data[token + 1]);
+    }
+
+    public void put(int hash, int index) {
+        if (hash == 0)
+            throw new IllegalArgumentException("hash must be nonzero");
+
+        int ptr = (hash & mask1) << 1;
+        int k = data[ptr];
+        if (k == 0) {
+            data[ptr] = hash;
+            data[ptr + 1] = index;
+            if (size >= resizeThreshold)
+                rehash(data.length * 2); //size is set inside
+            else
+                ++size;
+            return;
+        }
+        else if (k == hash && data[ptr + 1] == index) {
+            return;
+        }
+
+        while (true) {
+            ptr = (ptr + 2) & mask2; // next index
+            k = data[ptr];
+            if (k == 0) {
+                data[ptr] = hash;
+                data[ptr + 1] = index;
+                if (size >= resizeThreshold)
+                    rehash(data.length * 2); //size is set inside
+                else
+                    ++size;
+                return;
+            }
+            else if (k == hash && data[ptr + 1] == index) {
+                return;
+            }
+        }
+    }
+
+    public int size() {
+        return size;
+    }
+
+    private void rehash(final int newCapacity) {
+        resizeThreshold = (int) (newCapacity / 2 * fillFactor);
+        mask1 = newCapacity / 2 - 1;
+        mask2 = newCapacity - 1;
+
+        final int oldCapacity = data.length;
+        final int[] oldData = data;
+
+        data = new int[newCapacity];
+        size = 0;
+
+        for (int i = 0; i < oldCapacity; i += 2) {
+            final int oldHash = oldData[i];
+            if (oldHash != 0)
+                put(oldHash, oldData[i + 1]);
+        }
+    }
+
+    public static long nextPowerOfTwo( long x ) {
```

Comment:
If you like you can change the implementation to be:
```
x = -1 >>> Long.numberOfLeadingZeros(x - 1);
return x + 1;
```
See `ConcurrentHashMap`.

Reply:
`Long:numberOfLeadingZeros` implementation uses more method calls, conditional statements and binary operations. I would prefer to stick with the current implementation for its speed and simplicity.

Reply:
`numberOfLeadingZeros` is an intrinsic, so it will compile down to a machine instruction. Up to you.

Reply:
Oh, I've changed it, thanks.